base/reduce.jl

```diff
@@ -12,15 +12,18 @@ const SmallSigned = Union{Int8,Int16,Int32}
 const SmallUnsigned = Union{UInt8,UInt16,UInt32}
 end
 
-const CommonReduceResult = Union{UInt64,UInt128,Int64,Int128,Float32,Float64}
-const WidenReduceResult = Union{SmallSigned, SmallUnsigned, Float16}
+const CommonReduceResult = Union{UInt64,UInt128,Int64,Int128,Float16,Float32,Float64}
+const WidenReduceResult = Union{SmallSigned, SmallUnsigned}
 
+promote_sys_size{T}(::Type{T}) = T
+promote_sys_size{T<:SmallSigned}(::Type{T}) = Int
+promote_sys_size{T<:SmallUnsigned}(::Type{T}) = UInt
 # r_promote_type: promote T to the type of reduce(op, ::Array{T})
 # (some "extra" methods are required here to avoid ambiguity warnings)
 r_promote_type{T}(op, ::Type{T}) = T
-r_promote_type{T<:WidenReduceResult}(op, ::Type{T}) = widen(T)
-r_promote_type{T<:WidenReduceResult}(::typeof(+), ::Type{T}) = widen(T)
-r_promote_type{T<:WidenReduceResult}(::typeof(*), ::Type{T}) = widen(T)
+r_promote_type{T<:WidenReduceResult}(op, ::Type{T}) = promote_sys_size(T)
```

Comment:
As noted in #21523, widening is not necessarily appropriate for an arbitrary user-defined operator.   Maybe the default should be `T`.